#167972760 Setup and Integrate Transactional Email Service with Sendgrid

torsami77 opened the pull request:
#### Description of Task to be completed?
- User should be able to receive transactional email, user activities such as account verification, request for password recovery, trip request should trigger transactional email in order to aid user's experience

 #### How should this be manually tested?
- run ```git clone https://github.com/andela/wolfsbane-backend.git```
- cd into the cloned directory
- run ```git fetch origin pull/12/head:ft-integrate-email-service-167972760```
- run ```npm install``` : to install packages and ensure the @sendgrid and mocha libraries are installed
- Ensure your internet connection is stable
- Replace 'firstName' string from src/test/fakeData/fakeData.js file from with desired First Name
- Replace 'xxxxxx' string from src/test/fakeData/fakeData.js file with desired token
- Replace 'appropriatemail@gmail.com' string from src/test/fakeData/fakeData.js file with desired email
- The email category variable has been set for various tests already
- npm ```run test``` : to run mock test from src/test/services/autoMailer.test.js  and src/test/utils/emailTemplatesFunction.test.js

 #### Any background context you want to provide?
- To employ this service, import/ destructure senMail from services/index
- Ensure to have the SENDGRID_API_KEY variable set in your environment
- Ensure to pass recipient's email, category of email and provisional data to the email service function e.g: sendMail(email, confirmAccount, [firstName, token]);
- Email could be a string or array, category of email for now could be either of 'confirmAccount', 'passwordRecovery', or tripRequest, and the provisional data of course an array

 #### What are the relevant pivotal tracker stories?
[#167972760](https://www.pivotaltracker.com/story/show/167972760)

 #### Screenshots (if appropriate)
![001](https://user-images.githubusercontent.com/42122976/64324596-90b49380-cfbe-11e9-9bb3-e52d358f30ec.png)
![002](https://user-images.githubusercontent.com/42122976/64324601-914d2a00-cfbe-11e9-875a-2fd8e3a3fc6f.png)
![new trip request](https://user-images.githubusercontent.com/42122976/64324405-3ddadc00-cfbe-11e9-9fbb-84a2c0881d23.png)
![feedback101](https://user-images.githubusercontent.com/42122976/64324416-46331700-cfbe-11e9-9388-304bd35e6df1.png)


 #### Questions:

feobaby reviewed (commented):
Please squash your commits into one, _are you implying that we can only test this PR through unit testing?_
`Ensure to pass receiver's email and category of email` where and how?

Oliver-ke reviewed (changes requested):
Hi @torsami77, please squash your commit into one  and also be more descriptive on how we can test it manually

torsami77 commented:
Thanks so much Oliver

torsami77 commented:
@funmi5, please could you review the Pull Request again, let me know how to make it better

feobaby reviewed (dismissed):
I still don't understand this PR, please provide more insights into it. Also, you have a conflict,

uthdev reviewed (commented):
Nice work @torsami77. I really like how you separated the templates from the sendGrid services itself. I have tested your PR but the `Send Email Test`s are failing. Also, I expected that when I put my email address as directed by PR's how this should be tested, I should get emails sent from the app mailing services. Please, could you explain why the tests might be failing and why I didn't receive any emails?

darasimiolaifa commented on src/utils/emailTemplates/confirmAccountTemplate.js:
```diff
@@ -0,0 +1,246 @@
+/* eslint-disable no-useless-escape */
+let firstName, token;
```
Well done @torsami77. Good job so far. I'll just like to ask, why are you using global variables here? Is it completely unavoidable? I might be wrong, but I really avoid global variables in Javascript like a plague, unless they are absolutely necessary. Is it possible to have `theEmail` be a function that accepts `firstName` and `token`, and returns that template you are already assigning to it? That way, you can call `theEmail` from `confirmAccountTemplate`, and pass it the destructured `firstName` and `token` values, instead of declaring them as global variables. Please look into this.

torsami77 replied:
Sure it is possible to avoid global variable in this case and that would trigger a more than 25 lines concern for a function. At this time am thinking it would be better to have a global variable than a 200+ lines function. Please let me have your thoughts on the issue

darasimiolaifa replied:
Oh. Alright.

darasimiolaifa commented on src/services/autoMailer.js:
```diff
@@ -0,0 +1,31 @@
+    if (result[0] && result[0].request) {
+      return true;
+    }
+  } catch (error) {
```
Also, like @uthdev said, the tests for sending the email are failing. So I did a few console.log statements to trace the error source, and I found out that the `sendEmail` function returns a `401 unauthorized` error here. I tried changing the `from` email, maybe it was because that wasn't the email you registered with, but I got the same error. My other guess is probably we got the wrong API KEY. Can you help us look into this and get back? Thanks.

eokwukwe replied:
This conversation was marked as resolved, please I will need the information of how it was resolved

torsami77 commented:
Sure guys, that would be the API key, I once shared it on slack, my PC display is faulty and am posting this response from a phone that won't load slack, let me know if it is wise to paste it here so I can copy it from travisCI. I hope to attach my PC to external display by Sunday afternoon

darasimiolaifa commented:
Hi @torsami77 . I am about to review your PR, but please remove the `work in progress` label. Thanks.

darasimiolaifa commented:
And all the tests pass, and I get the email in my account. Well done, good job. I just think the template might need some improvements, but as touching the functionality, I am satisfied. *Thumbs up*

torsami77 commented:
Let me give the templates little more tweak @darasimiolaifa

feobaby reviewed (dismissed):
Good job Samson @Torsami

eokwukwe commented on src/services/autoMailer.js:
```diff
@@ -0,0 +1,31 @@
+
+sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+
+const sendEmail = async (receiver, category, data) => {
```
What is `receiver, category, data`. You need to add appropriate documentation to help us under what this function does and what the parameters are.

torsami77 replied:
@eokwukwe , since receiving notification about your feedback, all my effort to get electricity has been fruitless, though I have not giving up, I'm assured of having a generator fixed first thing tomorrow morning. I had to switch to a desktop computer as my laptop's display is still unstable at this time

eokwukwe replied:
Where is @uthdev?

eokwukwe reviewed (commented):
@torsami77 please include screenshots of the email to the PR

torsami77 committed:
feat(transactional email): setup and integrate sendgrid email service

- ensure user is able to receive transactional email
- ensure user is able to receive email verification upon signup

[Delivers #167972760]

eokwukwe reviewed (commented):
@torsami77 I need an explanation of what is going on with the content of this file `src/test/fakeData/fakeData.js`

eokwukwe commented on src/controllers/users.js:
```diff
@@ -2,9 +2,12 @@
 } from '../utils/index';
 
+
+let { baseUrl } = getCallbackUrls;
```
Why are you using `let` when you did not reassign the variable?

eokwukwe commented on src/controllers/users.js:
```diff
@@ -20,6 +23,7 @@ export default class UsersController {
    */
   static async registerUser(req, res) {
     try {
+      console.log(req.body);
```
Why is this `console.log` here?

eokwukwe commented on src/controllers/users.js:
```diff
@@ -33,6 +37,8 @@ export default class UsersController {
       const { id: userId } = user;
       const token = await generateToken({ userId });
+      const url = `${baseUrl}/confirmAccount/${userId}?token=${token}`;
+      await sendMail(email, 'confirmAccount', [req.body.firstName, url]);
```
You're using `sendEmail` here but it's not imported in the file

eokwukwe commented on src/controllers/resetPassword.js:
```diff
@@ -29,11 +31,13 @@ export default class ResetPasswordController {
-      const url = `${req.headers.host}/api/v1/resetpassword/${userId}?token=${token}`;
-      // await sendMail(email, 'passwordRecovery', url);
+      const url = `${baseUrl}/forgotpassword/${userId}?token=${token}`;
+      await sendMail(email, 'passwordRecovery', {firstName, url});
```
Is this `sendMail` or `sendEmail`? You're using `sendMail` and `sendEmail`. Please stick to one.

sa-ma commented on package.json:
```diff
@@ -58,5 +59,22 @@
     "sinon": "^7.4.1",
     "sinon-chai": "^3.3.0"
+  },
+  "nyc": {
```
Why do we have this here?

uthdev replied:
I have no idea. @torsami77  must have put it there.

torsami77 replied:
Yes @sa-ma , the nyc config helps direct nyc on files and folders to include and exclude when calculating the code coverage percentage, as we have redundant lines of html templates that could reduce code coverage

sa-ma replied:
I think that's why we have `codeclimate.yml`
